### PR TITLE
Add the ability to disable diagnostic on a request

### LINKED ISSUE #1173: Diagnostics Summary Visitor
CosmosDiagnostics should have a summarizer visitor that generates the summary instead of maintaining one along with the raw request level information.

### PR DESCRIPTION
**Pull request by j82w**
# Pull Request Template

## Description

Add a internal request option to disable diagnostic collection.
1. Refactored CosmosDiagnosticContext into an abstract class
2. CosmosDiagnosticContextSummary and CosmosDiagnosticContextList are merged into CosmosDiagnosticContextCore
3. Created CosmosDiagnosticContextCore and CosmosDiagnosticContextDisabled with a factory to choose diagnostic context is used based on the request options.

closes #1173

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Diagnostics/CosmosDiagnosticScope.cs [hunk 1]
```diff
@@ -47,7 +47,7 @@ public void Dispose()
         {
             if (this.isDisposed)
             {
-                throw new InvalidOperationException("The CosmosDiagnosticScope has already been disposed.");
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Diagnostics/CosmosDiagnosticScope.cs]: Any specific reason to change this behavior now?

File: Microsoft.Azure.Cosmos/src/Diagnostics/CosmosDiagnosticsContextDisabled.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+    using Microsoft.Azure.Cosmos.Diagnostics;
+
+    /// <summary>
+    /// This represents a disabled diagnostics context. This is used when the diagnostics
+    /// should not be recorded to avoid the overhead of the data collection.
+    /// </summary>
+    internal sealed class CosmosDiagnosticsContextDisabled : CosmosDiagnosticsContext
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Diagnostics/CosmosDiagnosticsContextDisabled.cs]: >CosmosDiagnosticsContextDisabled [](start = 26, length = 32)

Name seems very odd. 
Will it not collect any or will collect minimal retry context (Compute can leverage)?

File: Microsoft.Azure.Cosmos/src/Diagnostics/CosmosDiagnosticScope.cs (hunk 1 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Diagnostics/CosmosDiagnosticScope.cs]: This allows for a singleton instance to be reused for scenarios where the diagnostics does not want to be collected. This avoid possible null ref exceptions.

File: Microsoft.Azure.Cosmos/src/Diagnostics/CosmosDiagnosticsContextDisabled.cs (hunk 2 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Diagnostics/CosmosDiagnosticsContextDisabled.cs]: I agree, but I couldn't think of a better name at the time. How about `CosmosDiagnosticsContextEmpty`?

File: Microsoft.Azure.Cosmos/src/Diagnostics/CosmosDiagnosticScope.cs (hunk 1 above)
Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/Diagnostics/CosmosDiagnosticScope.cs]: Not fully on-board with this model but commiting 

---
In reply to: [373017149](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/1178#discussion_r373017149) [](ancestors = 373017149)

File: Microsoft.Azure.Cosmos/src/Diagnostics/CosmosDiagnosticsContextDisabled.cs (hunk 2 above)
Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/Diagnostics/CosmosDiagnosticsContextDisabled.cs]: Empty/None as Prefix looks better.

---
In reply to: [373018692](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/1178#discussion_r373018692) [](ancestors = 373018692)

**Pull request merged**